### PR TITLE
feat(rust, python): raise if window size in rolling functions isn't strictly positive

### LINKED ISSUE #9464: rolling_* functions inconsistent/wrong for period='0d'
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

Empty `period` results are inconsistent

The code below outputs
```python
shape: (1, 1)
┌────────┐
│ values │
│ ---    │
│ i64    │
╞════════╡
│ null   │
└────────┘
shape: (1, 1)
┌────────┐
│ values │
│ ---    │
│ i64    │
╞════════╡
│ 1      │
└────────┘
shape: (1, 1)
┌────────┐
│ values │
│ ---    │
│ i64    │
╞════════╡
│ 1      │
└────────┘
shape: (1, 1)
┌────────┐
│ values │
│ ---    │
│ i64    │
╞════════╡
│ null   │
└────────┘
```

### Reproducible example

```python
df = pl.DataFrame({'ts': [datetime(2020, 1, 1)], 'values': [1]}).sort('ts')
print(df.select(pl.col('values').rolling_sum(by='ts', window_size='0d', closed='left')))
print(df.select(pl.col('values').rolling_sum(by='ts', window_size='0d', closed='right')))
print(df.select(pl.col('values').rolling_sum(by='ts', window_size='0d', closed='both')))
print(df.select(pl.col('values').rolling_sum(by='ts', window_size='0d', closed='none')))
```

### Expected behavior

The window here, in the different cases, is:
- [2020-01-01, 2020-01-01)
- (2020-01-01, 2020-01-01]
- [2020-01-01, 2020-01-01]
- (2020-01-01, 2020-01-01)

Only the third one is non-empty, so I'd expect
```python

```python
shape: (1, 1)
┌────────┐
│ values │
│ ---    │
│ i64    │
╞════════╡
│ null   │
└────────┘
shape: (1, 1)
┌────────┐
│ values │
│ ---    │
│ i64    │
╞════════╡
│ null   │
└────────┘
shape: (1, 1)
┌────────┐
│ values │
│ ---    │
│ i64    │
╞════════╡
│ 1      │
└────────┘
shape: (1, 1)
┌────────┐
│ values │
│ ---    │
│ i64    │
╞════════╡
│ null   │
└────────┘
```

The refactor in https://github.com/pola-rs/polars/pull/9463/files made this a lot clearer

### Installed versions

<details>

```
--------Version info---------
Polars:      0.18.3
Index type:  UInt32
Platform:    Linux-5.10.102.1-microsoft-standard-WSL2-x86_64-with-glibc2.35
Python:      3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]

----Optional dependencies----
numpy:       1.24.3
pandas:      2.0.2
pyarrow:     12.0.0
connectorx:  <not installed>
deltalake:   0.9.0
fsspec:      <not installed>
matplotlib:  3.7.1
xlsx2csv:    <not installed>
xlsxwriter:  <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #9464

### CODE REVIEW

File: polars/polars-time/src/windows/groupby.rs [hunk 1]
```diff
@@ -461,38 +461,20 @@ pub(crate) fn groupby_values_iter<'a>(
...
-            time,
-            closed_window,
-            tu,
-            tz,
-            0,
-            None,
-        );
+    } else if matches!(closed_window, ClosedWindow::Both) {
+        // window only contains t
+        let iter = time.iter().enumerate().map(|(i, _)| Ok((i as IdxSize, 1)));
```
Comment (ritchie46) [inline polars/polars-time/src/windows/groupby.rs:473]: We don't have to iterator `time` in that case. It is cheaper to iterate `0..time.len()`

File: polars/polars-time/src/windows/groupby.rs [hunk 2]
```diff
@@ -461,38 +461,20 @@ pub(crate) fn groupby_values_iter<'a>(
...
-            period,
-            offset,
-            time,
-            closed_window,
-            tu,
-            tz,
-            0,
-            None,
-        );
+        // empty window
```
Comment (ritchie46) [inline polars/polars-time/src/windows/groupby.rs:476]: idem

File: polars/polars-time/src/windows/groupby.rs [hunk 3]
```diff
@@ -465,16 +465,19 @@ pub(crate) fn groupby_values_iter<'a>(
...
         let mut offset = period;
         offset.negative = true;
-        let iter =
-            groupby_values_iter_lookbehind(period, offset, time, closed_window, tu, tz, 0);
+        let iter = groupby_values_iter_lookbehind(period, offset, time, closed_window, tu, tz, 0);
         Box::new(iter)
     } else if matches!(closed_window, ClosedWindow::Both) {
         // window only contains t
-        let iter = time.iter().enumerate().map(|(i, _)| Ok((i as IdxSize, 1)));
+        let iter = (0..time.len())
```
Comment (ritchie46) [inline polars/polars-time/src/windows/groupby.rs:472]: Almost. Now we iterate from `0..len`, we don't need to `enumerate` (which also iterates from `0..len`).

So only: 

```rust
        let iter = (0..time.len())
            .map(|i| Ok((i as IdxSize, 0)));
```

Comment (MarcoGorelli) [reply polars/polars-time/src/windows/groupby.rs]: I should've thought of that - thanks, have updated!

**Comment by ritchie46** [comment]
> how about just raising if period isn't strictly positive? 

Even better!

File: polars/polars-time/src/windows/groupby.rs [hunk 4]
```diff
@@ -462,38 +462,10 @@ pub(crate) fn groupby_values_iter<'a>(
...
-            tz,
-            0,
-            None,
-        );
-        Box::new(iter)
-    }
+    offset.negative = true;
+    // t is at the right endpoint of the window
+    let iter = groupby_values_iter_lookbehind(period, offset, time, closed_window, tu, tz, 0);
+    Box::new(iter)
```
Comment (MarcoGorelli) [inline polars/polars-time/src/windows/groupby.rs:468]: so satisfying 😋

File: polars/polars-time/src/windows/groupby.rs [hunk 5]
```diff
@@ -595,7 +567,7 @@ pub fn groupby_values(
             Ok(flatten_par(&vals))
         })
     } else {
-        // Duration is 0 and window is closed on the left:
+        // Offset is 0 and window is closed on the left:
```
Comment (MarcoGorelli) [inline polars/polars-time/src/windows/groupby.rs:570]: drive-by typo fixing

File: polars/polars-time/src/windows/groupby.rs (hunk 4 above)
Comment (ritchie46) [reply polars/polars-time/src/windows/groupby.rs:468]: So all this didn't hit any test? Just to be sure, windows may still be lagging right?

Comment (MarcoGorelli) [reply polars/polars-time/src/windows/groupby.rs:468]: There are three branches here:

- if period > 0
- if period == 0 and closed is right or both
- if period == 0 and closed is left or none

if `period == 0` is disallowed, then only the first one is necessary, and the `if` statement can be removed and the code detended

`offset` and `period` are the same here, they just differ in sign, because the `rolling_*` functions don't have an `offset` argument. Windows where `offset != period` (like, that for point `2022-01-05` the window is `[2022-01-03, 2022-01-04)`) are only possible for `groupby_rolling` (and that's unaffected by this PR)

Comment (ritchie46) [reply polars/polars-time/src/windows/groupby.rs:468]: Thanks for explanation. Sorry for the questions on my side. I am working on many things, so I try to get an understanding without getting fully in the code. Looks good!

Comment (MarcoGorelli) [reply polars/polars-time/src/windows/groupby.rs:468]: Thanks! And please always do ask for clarifications when things aren't clear, this'll also help anyone in the future who ends up here from a `git blame` trying to figure out why a certain change was made
